File: schedule.py
import json
import constants as c
from datetime import datetime

class bowlingSchedule:
  def __init__(self):

    # Now lets open the season schedule for reference
    json_scedule = open(c.SCHEDULE_PATH, "r")
    self.schedule = json.load(json_scedule)
    
  def getCurrentWeekNumbers(self):
    weekNumbers = []
    
    # Loop over all the weeks in the schedule
    for week in self.schedule["weeks"]:
      # If the week in question happend today or in the past, we will try to scrape it
      date = datetime.strptime(week["date"], '%m/%d/%Y')
      if date < datetime.now():
        weekNumbers.append(int(week["weekNum"]))
    
    return weekNumbers
  
  def getWeek(self, weekNum):
    # Loop over all the weeks in the schedule
    for week in self.schedule["weeks"]:
      if int(week["weekNum"]) == int(weekNum):
        return week
    return None
  
  def getTimestamp(self, weekNum):
    dateStr = self.getWeek(weekNum)['date']
    return int(datetime.strptime(dateStr, "%m/%d/%Y").timestamp())
  
  def getLaneCount(self, weekNum = 1):
    return len(self.getWeek(weekNum)['matchups'])
  
  def getLaneNumber(self, weekNum, teamNum):
    week = self.getWeek(weekNum)
    # Flatten the 2d matchup array
    lanes = [item for sublist in week['matchups'] for item in sublist]
    # We always start on lane 1
    return lanes.index(teamNum) + 1
  
  def getOpponentNumber(self, weekNum, teamNum):
    week = self.getWeek(weekNum)
    for matchup in week['matchups']:
      try:
        return matchup[0 if matchup.index(teamNum) else 1]
      except:
        pass

File: www/js/schedule.js
const MS_PER_SEC = 1000;
const MS_PER_MIN = MS_PER_SEC * 60;
const MS_PER_HOUR = MS_PER_MIN * 60;
const MS_PER_DAY = MS_PER_HOUR * 24;

class schedule {
    schedule = null;

    constructor () {
        return new Promise(resolve => {
            makeRequest("GET", SCHEDULE_PATH).then(responseText => {
                this.schedule = JSON.parse(responseText).weeks;
                resolve(this);
            });
        });
    }

    getCurrentWeekNumbers() {
        return this.schedule.filter(week => new Date(week.date) <= Date.now()).map(week => week.weekNum);
    }
    
    getCurrentSchedule() {
        let weeks = this.schedule.filter(week => new Date(week.date) >= (Date.now() - MS_PER_DAY)).map(week => week.weekNum);
        return (weeks.length) ? this.getWeek(weeks[0]) : undefined;
    }
    
    getWeek(weekNum) {
        return this.schedule.find(week => week.weekNum == weekNum);
    }

    getDate(weekNum) {
        return this.getWeek(weekNum).date;
    }

    getTimestamp(weekNum) {
        return (new Date(this.getDate(weekNum))).getTime();
    }

    getLaneCount(weekNum = 1) {
        return this.getWeek(weekNum).matchups.length * 2; 
    }

    getLaneNumber(weekNum, teamNum) {
        // We always start on lane 1, so add 1 to the result
        return this.getWeek(weekNum).matchups.flat().indexOf(teamNum) + 1;
    }

    getOpponentNumber(weekNum, teamNum) {
        let matchup = this.getWeek(weekNum).matchups.find(matchup => matchup.includes(teamNum));
        return matchup[(matchup.indexOf(teamNum)) ? 0 : 1];
    }
}